### PR TITLE
Add fstatfs detour

### LINKED ISSUE #1270: Out of memory failure when retrieving >1GB file
### Bug Description

When the application tries to retrieve the [grype](https://github.com/anchore/grype) database, which is a >1GB sqlite file located on the pod, it crashes with an out of memory error. 

The database is retrieved with this code:
```
import (
"github.com/anchore/grype/grype"
"github.com/anchore/grype/grype/db"
)

conf := db.Config{
		DBRootDir:  dbRootDir,
		ListingURL: "",
	}
grypeDB,_, _, err := grype.LoadVulnerabilityDB(conf, false)
```
mirrord version: 3.33.0

Logs:
[agent.log](https://github.com/metalbear-co/mirrord/files/11190456/agent.log)
[layer.log](https://github.com/metalbear-co/mirrord/files/11190457/layer.log)

Notes:

- Increasing Goland's allocated memory heap didn't work
- Changing mirrord's configuration to read the file locally worked

### Steps to Reproduce

See description.

### Backtrace

_No response_

### Relevant Logs

_No response_

### Your operating system and version

MacOS

### Local process

Go

### Local process version

_No response_

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by DmitryDodzin**
Add support for `fstatfs`

closes #1270

### CODE REVIEW

File: mirrord/layer/src/file/hooks.rs [hunk 1]
```diff
@@ -503,6 +504,21 @@ unsafe extern "C" fn fill_stat(out_stat: *mut stat, metadata: &MetadataInternal)
...
+    let out = &mut *out_stat;
+    // on macOS the types might be different, so we try to cast and do our best..
+    out.f_type = best_effort_cast(metadata.r#type);
+    out.f_bsize = best_effort_cast(metadata.bsize);
+    out.f_blocks = metadata.blocks;
+    out.f_bfree = metadata.bfree;
+    out.f_bavail = metadata.bavail;
+    out.f_files = metadata.files;
+    out.f_ffree = metadata.ffree;
+    // out.f_fsid = metadata.fsid;
```
Comment (meowjesty) [inline mirrord/layer/src/file/hooks.rs:519]: Why is this commented?

File: mirrord/layer/src/file/hooks.rs [hunk 2]
```diff
@@ -585,6 +601,17 @@ unsafe extern "C" fn fstatat_detour(
         .unwrap_or_bypass_with(|_| FN_FSTATAT(fd, raw_path, out_stat, flag))
 }
 
+#[hook_guard_fn]
+unsafe extern "C" fn fstatfs_detour(fd: c_int, out_stat: *mut statfs) -> c_int {
```
Comment (meowjesty) [inline mirrord/layer/src/file/hooks.rs:605]: Shouldn't we check that `*mut statfs` is a valid pointer:

> [EFAULT]		The sb argument	points to an invalid address.

Comment (meowjesty) [reply mirrord/layer/src/file/hooks.rs:605]: Maybe implement [`CheckedInto`](https://metalbear-co.github.io/mirrord/mirrord_layer/common/trait.CheckedInto.html) for `*mut statfs`?

File: mirrord/layer/src/file/hooks.rs [hunk 3]
```diff
@@ -503,6 +504,21 @@ unsafe extern "C" fn fill_stat(out_stat: *mut stat, metadata: &MetadataInternal)
     out.st_blocks = best_effort_cast(metadata.blocks);
 }
 
+/// Fills the `stat` struct with the metadata
+unsafe extern "C" fn fill_statfs(out_stat: *mut statfs, metadata: &FsMetadataInternal) {
+    out_stat.write_bytes(0, 1);
```
Comment (meowjesty) [inline mirrord/layer/src/file/hooks.rs:509]: Add a comment telling why we need this `write_bytes`.

File: mirrord/agent/src/file.rs [hunk 4]
```diff
@@ -569,6 +573,31 @@ impl FileManager {
...
+            .open_files
+            .get(&fd)
+            .ok_or(ResponseError::NotFound(fd))?;
+
+        let statfs = match target {
+            RemoteFile::File(file) => nix::sys::statfs::fstatfs(file).unwrap(),
+            RemoteFile::Directory(path) => nix::sys::statfs::statfs(path).unwrap(),
+        };
+
+        let metadata = FsMetadataInternal {
```
Comment (meowjesty) [inline mirrord/agent/src/file.rs:588]: If this conversion can't fail, what do you think of moving it to a `From` implementation?

**Review by meowjesty — CHANGES_REQUESTED** [review]
Missing hooks are always fun.

File: mirrord/layer/src/file/hooks.rs (hunk 3 above)
Comment (DmitryDodzin) [reply mirrord/layer/src/file/hooks.rs:509]: TBH I'm not sure if it is needed I just started by copying `fill_stat`

Comment (DmitryDodzin) [reply mirrord/layer/src/file/hooks.rs:509]: It looks like it works without it too

**Review by meowjesty — APPROVED** [review]
:bear: :+1:

**Review by meowjesty — APPROVED** [review]
:bear: :+1:

**Pull request merged**